dev/Experimental/UnifiedDataModel.md

```diff
@@ -0,0 +1,326 @@
+        "instances": [
+            {"name": "m0", "model":"n", "fa_map":{"d": "da", "g": "ga", "s": "s", "b": "b"}, "parameters": {<params of mn2>} },
+            {"name": "m1", "model":"n", "fa_map":{"d": "db", "g": "gb", "s": "s", "b": "b"}, "parameters": {<params of mn3>} }
+        ]
+    },
+    {
+        "name": "<guid4>",
+        "pins": ["da", "db", "s"],
+        "constraints": [{"constraint": "Generator", "name": "mos",...}],
+        "instances": [
+            {"name": "m0", "model":"p", "fa_map":{...}, "parameters": {<params of mp4>} },
+            {"name": "m1", "model":"p", "fa_map":{...}, "parameters": {<params of mp5>} }
+        ]
+    }
+]
+```
+
+ALIGN calls the annotated generators from PDK to create primitives (no hard coded model names).
+
+__phase=2__ (After primitive generation, before placement)
+```python
+# Hierarchy
+[
+    {
+        "name": "ota_five",
+        "pins": [...],
+        "constraints": [
+            {"constraint": "Generator", "instances": "mn0", "parameters":{<param1>}},
+            {"constraint": "Generator", "instances": "mn1", "parameters":{<param2>}}
+        ],
+        "instances": [
+            {"name": "mn0", "model": "<guid1>", "fa_map": {...}},
+            {"name": "mn1", "model": "<guid2>", "fa_map": {...}},
+            {"name": "mn2_mn3", "model": "<guid3>", "fa_map": {...}}},
+            {"name": "mp4_mp5", "model": "<guid4>", "fa_map": {...}}},
+        ]
+    },
+    {
+        "name": "<guid1>",
+        "pins": ["d", "g", "s", "b"],
+        "constraints": [{"constraint": "Generator", "name": "mos", "instances": "m0", "parameters":{<param1>}}],
+        "instances": [{"name": "m0", "model":"n", "fa_map": {...}, "parameters": {<params of mn0>}}]
+    },
+    {
+        "name": "<guid2>",
+        "pins": ["d", "g", "s", "b"],
+        "constraints": [{"constraint": "Generator", "name": "mos", "instances": "m0", "parameters":{<param2>}}],
+        "instances": [{"name": "m0", "model":"n", "fa_map": {...}, "parameters": {<params of mn1>}}]
+    },
+    {
+        "name": "<guid3>",
+        "pins": ["da", "db", "ga", "gb", "s", "b"],
+        "constraints": [{"constraint": "Generator", "name": "mos",...}],
+        "instances": [
+            {"name": "m0", "model":"n", "fa_map":{"d": "da", "g": "ga", "s": "s", "b": "b"}, "parameters": {<params of mn2>} },
+            {"name": "m1", "model":"n", "fa_map":{"d": "db", "g": "gb", "s": "s", "b": "b"}, "parameters": {<params of mn3>} }
+        ]
+    },
+    {
+        "name": "<guid4>",
+        "pins": ["da", "db", "s"],
+        "constraints": [{"constraint": "Generator", "name": "mos",...}],
+        "instances": [
+            {"name": "m0", "model":"p", "fa_map":{...}, "parameters": {<params of mp4>} },
+            {"name": "m1", "model":"p", "fa_map":{...}, "parameters": {<params of mp5>} }
+        ]
+    }
+]
+# A separate data structure to store generated primitive variants 
```

Comment:
Why do we need a separate data structure? Can we use the same structure as a library?

Reply:
I am leaning against a separate data model as well @soneryaldiz as we will be routinely merging the two to query bbox data for checkers etc. We can think of caching things onto file once memory becomes a bottleneck.

Reply:
Thanks @parijatm . To make this object easy to inspect, efficient to manipulate and easier to code, could we use built-in Python types (list, dict, str) instead of our customized versions?